File: python_code/p006_medium_zigzag_conversion.py
# assume numRows>0


class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows < 2:
            return s
        # works when numRows>=2
        result = [[] for _ in range(numRows)]
        up_state, count = True, numRows
        for char in s:
            if up_state:
                result[numRows - count].append(char)
                count -= 1
                if count == 0:
                    up_state = False
                    count = 2
            else:
                result[-count].append(char)
                count += 1
                if count > numRows:
                    up_state = True
                    count = numRows - 1
        return ''.join([''.join(L) for L in result])

File: java_code/p006_medium_zigzag_conversion.java
// class Solution
class zigZagConversion {
    public String convert(String s, int numRows) {
        if (numRows < 2) return s;

        StringBuilder[] result = new StringBuilder[numRows];
        for (int i = 0; i < numRows; i++) {
            result[i] = new StringBuilder();
        }
        int count = 0;
        boolean up_state = true;
        int len = s.length();
        for (int i = 0; i < len; i++) {
            if (up_state) {
                result[count++].append(s.charAt(i));
                if (count >= numRows) {
                    up_state = false;
                    count--;
                }
            } else {
                result[--count].append(s.charAt(i));
                if (count == 0) {
                    up_state = true;
                    count = 1;
                }
            }
        }
        StringBuilder r = new StringBuilder();
        for (StringBuilder sb : result) {
            r.append(sb);
        }
        return r.toString();
    }

    public static void main(String[] args) {
        zigZagConversion z = new zigZagConversion();
        System.out.println("Test");
        System.out.println(z.convert("01210", 3));
    }
}